Applications/Spire/Source/KeyBindings/TaskKeysTableView.cpp

```diff
@@ -368,6 +383,144 @@ struct OrderTaskTableModel : TableModel {
   }
 };
 
+struct UniqueTaskKeyTableModel : TableModel {
+  static const auto KeyIndex = static_cast<int>(TableColumn::KEY);
+  static const auto RegionIndex = static_cast<int>(TableColumn::REGION);
+  using RegionKey = std::pair<Region, QKeySequence>;
+  std::shared_ptr<TableModel> m_source;
+  TableModelTransactionLog m_transaction;
+  std::unordered_map<RegionKey, int, RegionKeyHash> m_region_keys;
+  int m_conflicting_row;
+  scoped_connection m_source_connection;
+
+  explicit UniqueTaskKeyTableModel(std::shared_ptr<TableModel> source)
+      : m_source(std::move(source)),
+        m_conflicting_row(-1),
+        m_source_connection(m_source->connect_operation_signal(
+          std::bind_front(&UniqueTaskKeyTableModel::on_operation, this))) {
+    for(auto row = 0; row < get_row_size(); ++row) {
+      if(auto& key = get<QKeySequence>(row, KeyIndex); !key.isEmpty()) {
+        m_region_keys[{get<Region>(row, RegionIndex), key}] = row;
+      }
+    }
+  }
+
+  void update() {
+    if(m_conflicting_row != -1) {
+      auto& previous = get<QKeySequence>(m_conflicting_row, KeyIndex);
+      auto previous_row = m_conflicting_row;
+      m_conflicting_row = -1;
+      auto current_blocker = shared_connection_block(m_source_connection);
+      set(previous_row, KeyIndex, QKeySequence());
+      m_transaction.push(
+        UpdateOperation(previous_row, KeyIndex, previous, QKeySequence()));
+    }
+  }
+
+  int get_row_size() const override {
+    return m_source->get_row_size();
+  }
+
+  int get_column_size() const override {
+    return m_source->get_column_size();
+  }
+
+  AnyRef at(int row, int column) const override {
+    return m_source->at(row, column);
+  }
+
+  QValidator::State set(int row, int column, const std::any& value) override {
+    return m_source->set(row, column, value);
+  }
+
+  QValidator::State remove(int row) override {
+    return m_source->remove(row);
+  }
+
+  connection connect_operation_signal(
+      const OperationSignal::slot_type& slot) const override {
+    return m_transaction.connect_operation_signal(slot);
+  }
+
+  void on_operation(const Operation& operation) {
+    auto check_conflicting_row = [=] (const RegionKey& region_key) {
```

Comment:
I don't understand how this works.

My understanding is if the key or region column is updated, then a scan is performed on the table to find another row that has the same key and a region and resets the key for that row.

Why is all this additional work being performed? Realistically the implementation for this should be in the ```set``` method.

Reply:
Since resetting the key based on specs is performed when the user submits the region or key, rather than when modifying them.  The work in `on_operation` is to find the conflict row from the update of the region and key. The function `update` performs the reset when the user submits the update.

Reply:
Thanks for explaining this. Reimplement the model to reset in the ```set``` method instead of on submit. The specs will be updated to reflect that.